GW2EIEvtcParser/EncounterLogic/Fractals/SunquaPeak/AiKeeperOfThePeak.cs

```diff
@@ -400,12 +400,350 @@ protected override void SetInstanceBuffs(ParsedEvtcLog log)
             }
         }
 
+        private AgentItem GetAiAgentAt(long time)
+        {
+            AgentItem elementalAi = Targets.FirstOrDefault(x => x.IsSpecies(TargetID.AiKeeperOfThePeak))?.AgentItem;
+            AgentItem darkAi = Targets.FirstOrDefault(x => x.IsSpecies(TargetID.AiKeeperOfThePeak2))?.AgentItem;
+            if (elementalAi != null && elementalAi.InAwareTimes(time))
+            {
+                return elementalAi;
+            }
+            else if (darkAi != null && darkAi.InAwareTimes(time))
+            {
+                return darkAi;
+            }
+            return null;
+        }
+
         internal override void ComputePlayerCombatReplayActors(AbstractPlayer p, ParsedEvtcLog log, CombatReplay replay)
         {
             base.ComputePlayerCombatReplayActors(p, log, replay);
+
+            // tether between sprite and player
+            List<AbstractBuffEvent> spriteFixations = GetFilteredList(log.CombatData, new long[] { FixatedEnragedWaterSprite }, p, true, true);
+            replay.AddTether(spriteFixations, Colors.Purple, 0.5);
+
             // Tethering Players to Fears
             List<AbstractBuffEvent> fearFixations = GetFilteredList(log.CombatData, new long[] { FixatedFear1, FixatedFear2, FixatedFear3, FixatedFear4 }, p, true, true);
             replay.AddTether(fearFixations, Colors.Magenta, 0.5);
         }
+
+        internal override void ComputeNPCCombatReplayActors(NPC target, ParsedEvtcLog log, CombatReplay replay)
+        {
+            base.ComputeNPCCombatReplayActors(target, log, replay);
+
+
+            IReadOnlyList<AnimatedCastEvent> casts = log.CombatData.GetAnimatedCastData(target.AgentItem);
+            switch (target.ID)
+            {
+                case (int)TrashID.CCSorrowDemon:
+                    {
+                        const long fullCastDuration = 11840;
```

Comment:
The const here should be renamed to something that indicates that they are linked to Overwhelming sorrow